fitbenchmarking/parsing/tests/test_parsers.py

```diff
@@ -0,0 +1,235 @@
+                                         test_name_with_ext[0])
+            else:
+                raise RuntimeError(
+                    'Too many "{}" files found for "{}" test'.format(
+                        file_format, test_name))
+
+            test_parsers = {}
+            test_parsers['file_format'] = file_format
+            test_parsers['test_file'] = test_file
+            test_parsers['expected'] = load_expectation(expected_path)
+            params['test_parsers'].append(test_parsers)
+
+            test_factory = {}
+            test_factory['file_format'] = file_format
+            test_factory['test_file'] = test_file
+            params['test_factory'].append(test_factory)
+
+        func_eval = os.path.join(test_dir,
+                                 file_format,
+                                 'function_evaluations.json')
+        test_func_eval = {}
+        test_func_eval['file_format'] = file_format
+        test_func_eval['evaluations_file'] = func_eval
+        params['test_function_evaluation'].append(test_func_eval)
+
+    return params
+
+
+def load_expectation(filename):
+    """
+    Load an expected fitting problem from a json file.
+
+    :param filename: The path to the expectation file
+    :type filename: string
+    :return: A fitting problem to test against
+    :rtype: fitbenchmarking.parsing.FittingProblem
+    """
+    with open(filename, 'r') as f:
+        expectation_dict = load(f)
+
+    expectation = FittingProblem()
+    expectation.name = expectation_dict['name']
+    expectation.start_x = expectation_dict['start_x']
+    expectation.end_x = expectation_dict['end_x']
+    expectation.data_x = np.array(expectation_dict['data_x'])
+    expectation.data_y = np.array(expectation_dict['data_y'])
+    expectation.data_e = expectation_dict['data_e']
+    if expectation.data_e is not None:
+        expectation.data_e = np.array(expectation.data_e)
+    expectation.functions = expectation_dict['functions']
+    expectation.starting_values = expectation_dict['starting_values']
+    expectation.value_ranges = expectation_dict['value_ranges']
+
+    return expectation
+
+
+class TestParsers:
+    """
+    A class to hold the tests for parametrized testing of parsers.
+    """
+
+    params = generate_test_cases()
+
+    def test_parsers(self, file_format, test_file, expected):
+        """
+        Test that the parser correctly parses the specified input file.
+
+        :param file_format: The name of the file format
+        :type file_format: string
+        :param test_file: The path to the test file
+        :type test_file: string
+        """
+        assert (test_file is not None), \
+            'No test file for {}'.format(file_format)
+
+        # Test import
+        module = import_module(name='.{}_parser'.format(file_format),
+                               package='fitbenchmarking.parsing')
+
+        parser = getmembers(module, lambda m: (isclass(m)
```

Comment:
I appreciate that you have used introspection to determine which parsers to test in order to reduce the burden on contributors who are adding parsers; however I would say that the downside to this is it risks the possibility of tests passing because you are not testing some/all of the parsers.  This risk is essentially removed if you explicitly state the parsers to be test.

I think you should keep your current approach, but it is something to be aware of.  I would also say that it would probably be good if you could add something which tests that the base set of parsers (e.g. fitbenchmark, sasview, nist) are included in parsers determined bye introspection.  This way you won't have an additional contributor burden, but will be more confident that your introspection will be discovering all of the required parsers to test.

It might also be worth noting in the documentation for adding a new parser that the while the contributor doesn't have to modify test_parsers.py, they should run `pytest -vv fitbenchmarking/parsing/tests/test_parsers.py` to ensure that the new parser they have added has the required tests.

Reply:
Yes, good point. I'll update the docs and check for the ones we know about at the moment.